Modules/arraymodule.c

```diff
@@ -669,34 +805,176 @@ newarrayobject(PyTypeObject *type, Py_ssize_t size, const struct arraydescr *des
     return (PyObject *) op;
 }
 
+static inline int
+valid_index(Py_ssize_t i, Py_ssize_t limit)
+{
+    return (size_t) i < (size_t) limit;
+}
+
+static inline PyObject *
+getarrayitem(arrayobject *ap, Py_ssize_t i, char *items)
```

Comment:
I think this is going to be a bit cleaner and easier to understand if it takes a `arraydata *` instead of `char *items`.  That will avoid the `_Py_CONTAINER_OF()`, which can be a bit confusing to understand.